File: code/cpp/ch25/solutions/warmup_05_sol.cpp
/*
 * Solution for Warmup 5: Longest Common Subsequence
 * Chapter 25: Dynamic Programming III — Subsequences & Knapsack
 */
#include <algorithm>
#include <climits>
#include <iostream>
#include <string>
#include <vector>
using namespace std;

int solve(string text1, string text2) {
    int m = text1.size(), n = text2.size();
    vector<int> prev(n + 1, 0);
    for (int i = 1; i <= m; i++) {
        vector<int> curr(n + 1, 0);
        for (int j = 1; j <= n; j++) {
            if (text1[i-1] == text2[j-1]) curr[j] = prev[j-1] + 1;
            else curr[j] = max(prev[j], curr[j-1]);
        }
        prev = curr;
    }
    return prev[n];
}

int main() { return 0; }

File: code/python/ch25/solutions/warmup_05_sol.py
"""
Solution for Warmup 5: Longest Common Subsequence
====================================================
Chapter 25: Dynamic Programming III — Subsequences & Knapsack

APPROACH
--------
Space-optimized 2-row DP.
dp[j] = LCS length of text1[:i] and text2[:j].

TIME COMPLEXITY:  O(m * n) where m, n are string lengths
SPACE COMPLEXITY: O(n)
"""


def solve(text1: str, text2: str) -> int:
    """Return the length of the longest common subsequence."""
    m, n = len(text1), len(text2)
    prev = [0] * (n + 1)
    for i in range(1, m + 1):
        curr = [0] * (n + 1)
        for j in range(1, n + 1):
            if text1[i - 1] == text2[j - 1]:
                curr[j] = prev[j - 1] + 1
            else:
                curr[j] = max(prev[j], curr[j - 1])
        prev = curr
    return prev[n]


# ── Do not change anything below this line ──────────────────────
if __name__ == "__main__":
    text1 = input().strip()
    text2 = input().strip()
    print(solve(text1, text2))
